File: packages/exochain-sdk/src/transport/http.ts
// Copyright 2026 Exochain Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at:
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

/**
 * HTTP transport for the `exo-gateway` REST API.
 *
 * Uses the global `fetch` which is available in Node 18+ and all modern
 * browsers. No third-party HTTP library is used to keep the SDK dependency-
 * free.
 */

import { TransportError } from '../errors.js';
import type { HealthResponse } from '../types.js';
import {
  assertJsonObject,
  validateHealthResponse,
  type JsonObject,
} from '../validation.js';

/** Options for {@link HttpTransport}. */
export interface HttpTransportOptions {
  /** Optional API key sent as `Authorization: Bearer <apiKey>`. */
  readonly apiKey?: string;
  /** Request timeout in milliseconds. Defaults to 30_000. */
  readonly timeout?: number;
  /** Override the `fetch` implementation (useful for tests). */
  readonly fetch?: typeof fetch;
}

/** Small fetch wrapper that serializes and deserializes JSON bodies. */
export class HttpTransport {
  readonly #baseUrl: string;
  readonly #apiKey?: string;
  readonly #timeout: number;
  readonly #fetch: typeof fetch;

  constructor(baseUrl: string, opts?: HttpTransportOptions) {
    if (typeof baseUrl !== 'string' || baseUrl.length === 0) {
      throw new TransportError('baseUrl is required');
    }
    this.#baseUrl = baseUrl.replace(/\/+$/, '');
    if (opts?.apiKey !== undefined) this.#apiKey = opts.apiKey;
    this.#timeout = opts?.timeout ?? 30_000;
    const f = opts?.fetch ?? globalThis.fetch;
    if (typeof f !== 'function') {
      throw new TransportError('fetch is not available in this environment');
    }
    this.#fetch = f;
  }

  /** Gateway `/health` probe. */
  public async health(): Promise<HealthResponse> {
    return validateHealthResponse(await this.get('/health'));
  }

  /** Issue a GET and parse the JSON body as untrusted data. */
  public async get(path: string): Promise<unknown> {
    return this.request('GET', path);
  }

  /** Issue a POST with a JSON body and parse the JSON response as untrusted data. */
  public async post(path: string, body: JsonObject): Promise<unknown> {
    return this.request('POST', path, assertJsonObject(body, `${path} request body`));
  }

  async #abortable(): Promise<{ signal: AbortSignal; cancel: () => void }> {
    const ctrl = new AbortController();
    const id = setTimeout(() => ctrl.abort(new Error('request timed out')), this.#timeout);
    return { signal: ctrl.signal, cancel: () => clearTimeout(id) };
  }

  async request(method: string, path: string, body?: JsonObject): Promise<unknown> {
    const url = `${this.#baseUrl}${path.startsWith('/') ? path : `/${path}`}`;
    const headers: Record<string, string> = { accept: 'application/json' };
    if (this.#apiKey !== undefined) {
      headers.authorization = `Bearer ${this.#apiKey}`;
    }
    let serialized: string | undefined;
    if (body !== undefined) {
      headers['content-type'] = 'application/json';
      serialized = JSON.stringify(body);
    }

    const { signal, cancel } = await this.#abortable();
    let res: Response;
    try {
      const init: RequestInit = { method, headers, signal };
      if (serialized !== undefined) {
        init.body = serialized;
      }
      res = await this.#fetch(url, init);
    } catch (err) {
      cancel();
      throw new TransportError(`network error: ${stringifyError(err)}`, { cause: err });
    }
    cancel();

    const text = await res.text();
    if (!res.ok) {
      throw new TransportError(`HTTP ${res.status} ${res.statusText} for ${method} ${path}`, {
        status: res.status,
        body: text,
      });
    }
    if (text.length === 0) {
      return undefined;
    }
    try {
      const parsed: unknown = JSON.parse(text);
      return parsed;
    } catch (err) {
      throw new TransportError('failed to parse JSON response', {
        cause: err,
        body: text,
      });
    }
  }
}

function stringifyError(err: unknown): string {
  if (err instanceof Error) return err.message;
  return String(err);
}

File: packages/exochain-py/exochain/transport/http.py
# Copyright 2026 Exochain Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at:
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Async HTTP transport built on :mod:`httpx`."""

from __future__ import annotations

from types import TracebackType
from typing import Any

import httpx

from ..errors import TransportError

_DEFAULT_USER_AGENT = "exochain-py/0.1.0"


class HttpTransport:
    """Thin async wrapper around ``httpx.AsyncClient`` with typed error mapping.

    All network errors are surfaced as :class:`~exochain.errors.TransportError`
    with ``status`` + ``body`` preserved so callers can retry on 503/429 or
    branch on 401 without parsing exception strings. (A-061)

    ``timeout`` accepts either a plain float (seconds, applied to every phase)
    or a fully-configured ``httpx.Timeout`` for per-phase control (connect,
    read, write, pool). (A-061)
    """

    def __init__(
        self,
        base_url: str,
        *,
        api_key: str | None = None,
        timeout: float | httpx.Timeout = 30.0,
    ) -> None:
        headers: dict[str, str] = {"User-Agent": _DEFAULT_USER_AGENT}
        if api_key:
            headers["Authorization"] = f"Bearer {api_key}"
        self._client: httpx.AsyncClient = httpx.AsyncClient(
            base_url=base_url,
            headers=headers,
            timeout=timeout,
        )

    async def health(self) -> dict[str, Any]:
        """Call ``GET /health`` and return the decoded JSON body."""
        return await self.get("/health")

    async def get(self, path: str) -> dict[str, Any]:
        """Issue a ``GET`` request and return the decoded JSON body."""
        try:
            response = await self._client.get(path)
            response.raise_for_status()
            data = response.json()
        except httpx.HTTPStatusError as exc:
            raise TransportError(
                f"GET {path} failed: {exc}",
                status=exc.response.status_code,
                body=exc.response.text,
            ) from exc
        except httpx.HTTPError as exc:
            raise TransportError(f"GET {path} failed: {exc}") from exc
        if not isinstance(data, dict):
            raise TransportError(f"GET {path} did not return a JSON object")
        return data

    async def post(self, path: str, body: dict[str, Any]) -> dict[str, Any]:
        """Issue a ``POST`` with a JSON body and return the decoded JSON response."""
        try:
            response = await self._client.post(path, json=body)
            response.raise_for_status()
            data = response.json()
        except httpx.HTTPStatusError as exc:
            raise TransportError(
                f"POST {path} failed: {exc}",
                status=exc.response.status_code,
                body=exc.response.text,
            ) from exc
        except httpx.HTTPError as exc:
            raise TransportError(f"POST {path} failed: {exc}") from exc
        if not isinstance(data, dict):
            raise TransportError(f"POST {path} did not return a JSON object")
        return data

    async def close(self) -> None:
        """Close the underlying HTTP client."""
        await self._client.aclose()

    async def __aenter__(self) -> HttpTransport:
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        await self.close()


__all__ = ["HttpTransport"]
